llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -8883,6 +8883,26 @@ static std::optional<bool> isImpliedCondICmps(const ICmpInst *LHS,
   if (L0 == R0 && L1 == R1)
     return isImpliedCondMatchingOperands(LPred, RPred);
 
+  // It only really makes sense in the context of signed comparison for "X - Y
+  // must be positive if X >= Y and no overflow".
+  // Take SGT as an example:  L0:x > L1:y and C >= 0
+  //                      ==> R0:(x -nsw y) < R1:(-C) is false
+  if ((LPred == ICmpInst::ICMP_SGT || LPred == ICmpInst::ICMP_SGE) &&
+      match(R0, m_NSWSub(m_Specific(L0), m_Specific(L1)))) {
+    if (match(R1, m_NonPositive()) &&
+        isImpliedCondMatchingOperands(LPred, RPred) == false)
+      return false;
+  }
```

Comment:
Is this right for `sge` if `C == 0`? Your proofs checkout but afaict its b.c your proof is using `X != -X` to handle the false case which has problems with this case...

Reply:
Does  the link https://alive2.llvm.org/ce/z/N4uvTh proof it is valid for **sge** with **C == 0** ?

Reply:
I made a mistake think this is okay.